File: sherpout-server/src/main/java/com/sherpout/server/api/exercise/dto/ExerciseDTO.java
package com.sherpout.server.api.exercise.dto;

import com.sherpout.server.api.exercise.enumerated.ExerciseDifficulty;
import com.sherpout.server.api.exercise.enumerated.ExerciseEquipment;
import com.sherpout.server.api.exercise.enumerated.Muscle;
import com.sherpout.server.commons.dto.ImageDTO;
import com.sherpout.server.commons.dto.TranslatedStringDTO;
import com.sherpout.server.commons.validation.TranslatedStringValid;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Null;
import jakarta.validation.constraints.Size;
import lombok.Getter;
import lombok.Setter;

import java.util.List;
import java.util.Set;

@Getter
@Setter
public class ExerciseDTO {
    @Null
    private Long id;

    @NotNull
    @TranslatedStringValid
    private TranslatedStringDTO name;

    @NotNull
    @TranslatedStringValid
    private TranslatedStringDTO description;

    @NotNull
    private ExerciseDifficulty difficulty;

    @NotNull
    private Muscle targetMuscle;

    private Set<Muscle> supportMuscles;

    private Set<ExerciseEquipment> equipments;

    @Null
    private Integer likesNumber;

    private ImageDTO cover;

    private List<ImageDTO> images;
}

File: sherpout_mobile/lib/common/dto/exercise_dto.dart
import 'package:sherpoutmobile/common/dto/exercise_equipment.dart';
import 'package:sherpoutmobile/common/dto/exercise_list_dto.dart';
import 'package:sherpoutmobile/common/dto/translated_string_dto.dart';

import 'exercise_difficulty.dart';
import 'image_dto.dart';
import 'muscle.dart';

class ExerciseDto {
  final int id;
  final TranslatedStringDto name;
  final TranslatedStringDto description;
  final ExerciseDifficulty difficulty;
  final Muscle targetMuscle;
  final Set<Muscle>? supportMuscles;
  final Set<ExerciseEquipment>? equipments;
  final int likesNumber;
  final ImageDto? cover;
  final Set<ImageDto>? images;

  ExerciseDto({
    required this.id,
    required this.name,
    required this.description,
    required this.difficulty,
    required this.targetMuscle,
    this.supportMuscles,
    this.equipments,
    required this.likesNumber,
    this.cover,
    this.images
  });

  ExerciseListDto toListDto() {
    return ExerciseListDto(
        id: id,
        name: name,
        difficulty: difficulty,
        targetMuscle: targetMuscle,
        supportMuscles: supportMuscles,
        likesNumber: likesNumber,
        cover: cover
    );
  }

  factory ExerciseDto.fromJson(Map<String, dynamic> json) {
    return ExerciseDto(
        id: json['id'] as int,
        name: TranslatedStringDto.fromJson(json['name']),
        description: TranslatedStringDto.fromJson(json['description']),
        difficulty: ExerciseDifficulty.fromName(json['difficulty'])!,
        targetMuscle: Muscle.fromName(json['targetMuscle'])!,
        supportMuscles: (json['supportMuscles'] as List?)
            ?.map((e) => Muscle.fromName(e))
            .whereType<Muscle>()
            .toSet(),
        equipments: (json['equipments'] as List?)
            ?.map((e) => ExerciseEquipment.fromName(e))
            .whereType<ExerciseEquipment>()
            .toSet(),
        likesNumber: json['likesNumber'] as int,
        cover: json['cover'] != null ? ImageDto.fromJson(json['cover']) : null,
        images: (json['images'] as List?)
            ?.map((image) => ImageDto.fromJson(image))
            .whereType<ImageDto>()
            .toSet());
  }
}